### PR TITLE
Fix USE_AFTER_FREE (CWE-416)

### PR DESCRIPTION
**Pull request by FaramosCZ**
The "nonnul" variable will be freed on line 2756

### CODE REVIEW

**Comment by FaramosCZ** [comment]
Submited under MCA

**Comment by FaramosCZ** [comment]
```
Error: USE_AFTER_FREE (CWE-416):
mariadb-10.3.10/storage/innobase/row/row0merge.cc:2036: freed_arg: "free" frees "nonnull".
mariadb-10.3.10/storage/innobase/row/row0merge.cc:2761: double_free: Calling "free" frees pointer "nonnull" which has already been freed.
# 2759|         }
# 2760|         mem_heap_free(row_heap);
# 2761|->       ut_free(nonnull);
# 2762|   
# 2763|   all_done:
```

File: storage/innobase/row/row0merge.cc [hunk 1]
```diff
@@ -2029,7 +2029,6 @@ row_merge_read_clustered_index(
 					row = NULL;
 					mtr_commit(&mtr);
 					mem_heap_free(row_heap);
-					ut_free(nonnull);
 					goto write_buffers;
```
Comment (dr-m) [inline storage/innobase/row/row0merge.cc]: I do not think that this is correct. The alleged double free of `nonnull` is executed later in the code, but only if `row != NULL`. Here is the code from 10.0, which is the oldest applicable version:
```c++
		if (row == NULL) {
			goto all_done;
		}

		if (err != DB_SUCCESS) {
			goto func_exit;
		}
	}

func_exit:
	mtr_commit(&mtr);

	mem_heap_free(row_heap);

	if (nonnull) {
		mem_free(nonnull);
	}

all_done:
```

**Comment by dr-m** [comment]
OK, I see it now. The relevant part of the log is:
```
mariadb-10.3.10/storage/innobase/row/row0merge.cc:2036: freed_arg: "free" frees "nonnull".
…
mariadb-10.3.10/storage/innobase/row/row0merge.cc:2631: cond_true: Condition "row == NULL", taking true branch.
mariadb-10.3.10/storage/innobase/row/row0merge.cc:2631: cond_false: Condition "file->fd.operator os_file_t() == -1 /* OS_FILE_CLOSED */", taking false branch.
mariadb-10.3.10/storage/innobase/row/row0merge.cc:2667: else_branch: Reached else branch.
mariadb-10.3.10/storage/innobase/row/row0merge.cc:2668: cond_true: Condition "!row_merge_file_create_if_needed(file, tmpfd, buf->n_tuples, path)", taking true branch.
mariadb-10.3.10/storage/innobase/row/row0merge.cc:2673: goto: Jumping to label "func_exit".
mariadb-10.3.10/storage/innobase/row/row0merge.cc:2754: label: Reached label "func_exit".
mariadb-10.3.10/storage/innobase/row/row0merge.cc:2757: cond_true: Condition "mtr.is_active()", taking true branch.
mariadb-10.3.10/storage/innobase/row/row0merge.cc:2761: double_free: Calling "free" frees pointer "nonnull" which has already been freed.
```
I believe that this bug was introduced in MySQL 5.6.20 and 5.7.10 (commit f64bc45505dcd8ca59d52e4091f14fff092a37e5) and merged to MariaDB 10.0.

It seems to me that when `row==NULL` (we have reached the end of the input), maybe we should not be creating a file here at all. But for some reason, we already had created {{file->fd}}.

@Thirunarayanan, you are the original author of the change. Can you please analyze the above code flow? We could silence the warnings by assigning `nonnull=NULL`, but I’d like to ensure that this flow of execution does not indicate some other error.

**Comment by dr-m** [comment]
I took another look at these messages:
```
mariadb-10.3.10/storage/innobase/row/row0merge.cc:2036: freed_arg: "free" frees "nonnull".
…
mariadb-10.3.10/storage/innobase/row/row0merge.cc:2631: cond_true: Condition "row == NULL", taking true branch.
mariadb-10.3.10/storage/innobase/row/row0merge.cc:2631: cond_false: Condition "file->fd.operator os_file_t() == -1 /* OS_FILE_CLOSED */", taking false branch.
mariadb-10.3.10/storage/innobase/row/row0merge.cc:2667: else_branch: Reached else branch.
mariadb-10.3.10/storage/innobase/row/row0merge.cc:2668: cond_true: Condition "!row_merge_file_create_if_needed(file, tmpfd, buf->n_tuples, path)", taking true branch.
mariadb-10.3.10/storage/innobase/row/row0merge.cc:2673: goto: Jumping to label "func_exit".
```
If `file->fd == OS_FILE_CLOSED` in line 2631, then `row_merge_file_create_if_needed()` would invoke `row_merge_file_create()` and `row_merge_tmpfile_if_needed()`. If either one fails, the function will return `false`, and the condition in line 2668 would indeed hold, leading to the reported double-free.

The problem seems to be the `goto func_exit`. Before 10.2 (or MySQL 5.7), the error handling did `break` or `goto all_done`, which would avoid the double-freeing.

It seems to me that also the `row_heap` could be double-freed, but I would assume that the static analysis tool had not been configured to treat `mem_heap_create()` and `mem_heap_free()` as allocation functions.

The logic of the function has become rather convoluted over time due to the deferred creation of the temporary files, and the introduction of `BtrBulk` and some code for spatial index creation.

**Pull request closed without merging**